### PR TITLE
Fix error handling for Register

### PR DESCRIPTION
**Pull request by jamesmaa**
I looked into the koa docs and found some examples I was modeling off of. It seems to produce simpler and more readable code and it works.

https://github.com/koajs/koa#koa-v2
http://stackoverflow.com/questions/22059775/using-callbacks-with-nodejs-in-koa

The problem with the Login error handling fix is that ultimately it is convoluted to differentiate between different types of errors, and requires additional JSON parsing.

Happy to discuss whether this is the right approach or not

👓 @phil303

### CODE REVIEW

**Comment by phil303** [comment]
@jamesmaa Regarding the error handling of the Login patch - while I agree that it's not ideal, I'm not sure there's another way. Even in this patch, you're doing type checking. It seems like it'd be more explicit to define errors into a "class" of error and check for that.

**Comment by phil303** [comment]
Thanks for this patch! Looks good but let's chat about the error handling stuff :nail_care:

**Comment by jamesmaa** [comment]
💇 Adding @schwers because I made a change to Login error handling

File: src/app/router/handlers/Login.js [hunk 1]
```diff
@@ -24,7 +24,9 @@ export default class Login extends BaseHandler {
 
     } catch (e) {
       if (e instanceof errors.ValidationError) {
-        dispatch(sessionActions.sessionError('ERROR NOT USED'));
+        if (e.errors && e.errors[0]) {
```
Comment (phil303) [inline src/app/router/handlers/Login.js]: for this line and the similar one in Register, can you move the checks into the conditional above.
```js
if (e instanceof errors.ValidationError && e.errors && e.errors[0]) {
  dispatch(sessionActions.sessionError(e.errors[0].error));
}
```

File: src/server/session/registerproxy.js [hunk 2]
```diff
@@ -11,16 +14,20 @@ export default (router, apiOptions) => {
...
+      ctx.status = 401;
+      if (error && (typeof error === 'string')) {
+        ctx.type = 'application/json';
+        ctx.body = { error };
+      } else { // we're not sure what this error is, log it for now
+        errorLog({
+          error,
+          requestUrl: ctx.request.url,
+          userAgent: ctx.headers['user-agent'],
+        },{
```
Comment (phil303) [inline src/server/session/registerproxy.js]: nit, space between your comma and your last curly

**Comment by phil303** [comment]
:nail_care: small comments and give a chance for @schwers to weigh in

File: src/server/session/registerproxy.js [hunk 3]
```diff
@@ -11,16 +14,20 @@ export default (router, apiOptions) => {
...
-    } catch (e) {
-      console.log(e);
-      let errormsg = 'UNKNOWN_ERROR';
-
-      if (e && (typeof e === 'string')) {
-        errormsg = e;
+    } catch (error) {
+      ctx.status = 401;
+      if (error && (typeof error === 'string')) {
+        ctx.type = 'application/json';
```
Comment (schwers) [inline src/server/session/registerproxy.js]: 🔕  I think Koa will do this for you automatically if you set body to a dictionary like you're doing in the line below.

File: src/server/session/registerproxy.js [hunk 4]
```diff
@@ -11,16 +14,20 @@ export default (router, apiOptions) => {
...
+        ctx.body = { error };
+      } else { // we're not sure what this error is, log it for now
+        errorLog({
```
Comment (schwers) [inline src/server/session/registerproxy.js]: 🔕  If we keep doing this we should make a helper for it, something like `logServerError(ctx)`

**Comment by schwers** [comment]
💅 agree with Phil's comments, left a few comments on non-blocking stuff, overall patch looks good to me.

**Comment by jamesmaa** [comment]
💇

File: src/lib/errorLog.js [hunk 5]
```diff
@@ -173,6 +174,16 @@ export const formatLogJSON = logJSON => {
   return JSON.stringify(logJSON, null, 2).replace(/\\n/g, '\n');
 };
 
+export const logServerError = (error, ctx) => {
```
Comment (schwers) [inline src/lib/errorLog.js]: Thanks for adding the helper! Could you change https://github.com/reddit/reddit-mobile/blob/2X/src/server/meta/stats.js#L40 to use it too?

**Comment by phil303** [comment]
:fish: thanks!

**Comment by schwers** [comment]
The helper is nice, left a comment on one other place we should use it. I'll optimistically :fish: but make sure @phil303 is on-board too

**Comment by jamesmaa** [comment]
7c1c92c5d0cea472168c2fad69efc157b17f05f4

**Pull request closed without merging**